### PR TITLE
feat: nominate beneficiary holder

### PR DESCRIPTION
**Pull request by simboonlong**
### pr updates
- beneficiary can nominate new beneficiary and holder

### user story
- https://www.pivotaltracker.com/story/show/173008554

### preview
Screenshot 2020-06-03 at 11 08 20 AM
Screenshot 2020-06-03 at 2 38 21 PM

### CODE REVIEW

File: src/components/AssetManagementPanel/AssetManagementForm/AssetManagementForm.tsx [hunk 1]
```diff
@@ -32,24 +36,34 @@ export const AssetManagementForm = ({
...
 }: AssetManagementFormProps) => {
   const isHolder = account === holder;
   const isBeneficiary = account === beneficiary;
   const canSurrender = isBeneficiary && isHolder;
   const canEndorseBeneficiary = isBeneficiary && isHolder;
+  const canNominateBeneficiaryHolder =
+    isBeneficiary &&
+    !isHolder &&
+    approvedBeneficiary === "0x0000000000000000000000000000000000000000" &&
+    approvedHolder === "0x0000000000000000000000000000000000000000";
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/AssetManagementForm.tsx]: Dont need any of these conditions, the beneficiary can nominate new ones any. For instance if the holder do not agree to the values, he can ask the bene to re-nominate.

File: src/components/AssetManagementPanel/AssetManagementForm/AssetManagementForm.tsx [hunk 2]
```diff
@@ -70,6 +84,20 @@ export const AssetManagementForm = ({
...
 
+    case AssetManagementActions.NominateBeneficiaryHolder:
+      return (
+        <NominateBeneficiaryHolderForm
+          formAction={formAction}
+          tokenId={tokenId}
+          tokenRegistryAddress={tokenRegistryAddress}
+          beneficiary={beneficiary}
+          holder={holder}
+          handleTransfer={onApproveNewTransferTargets}
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/AssetManagementForm.tsx]: `handleTransfer` is a weird action name for this form

File: src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.test.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+      const beneficiaryComponent = container.getByTestId("editable-input-beneficiary");
+      const holderComponent = container.getByTestId("editable-input-holder");
+
+      expect(beneficiaryComponent).not.toBeNull();
+      expect(holderComponent).not.toBeNull();
+    });
+  });
+
+  it("should have the endorse button and cancel button", async () => {
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.test.tsx]: is it still the `endorse` button?

File: src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      showOverlay(
+        showDocumentTransferMessage(MessageTitle.NOMINATE_BENEFICIARY_HOLDER_SUCCESS, {
+          isSuccess: true,
+        })
+      );
+    }
+  }, [isConfirmed, newBeneficiary, showOverlay]);
+
+  const isValidEndorse = () => {
+    if (!newBeneficiary || !newHolder) return false;
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.tsx:56]: Might also want to consider a function that checks if its a valid smart contract address. 
I think the `web3-utils` package does that with `isAddress` (https://www.npmjs.com/package/web3-utils).

Comment (yehjxraymond) [reply src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.tsx:56]: While at it, might want to add that condition for other forms as well lol

File: src/components/UI/Overlay/OverlayContent/DocumentTransferMessage.tsx [hunk 5]
```diff
@@ -138,6 +139,14 @@ export const MessageHolderSuccess = ({ address }: MessageProps) => {
   );
 };
 
+export const MessageNominateBeneficiaryHolderSuccess = () => {
+  return (
+    <>
+      <p>Document has been nominated successfully. Please await holder to execute transfer approval.</p>
```
Comment (yehjxraymond) [inline src/components/UI/Overlay/OverlayContent/DocumentTransferMessage.tsx]:
```suggestion
      <p>Document has been nominated successfully. Please notify holder to execute transfer.</p>
```

File: src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.tsx (hunk 4 above)
Comment (simboonlong) [reply src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.tsx:56]: yup okay

**Comment by yehjxraymond** [comment]
image

Verified that the state is successfully updated on the blockchain:

https://ropsten.etherscan.io/address/0x37531cfb40f82dc223dd7c93957d384729db8183#readContract

File: src/components/AssetManagementPanel/AssetManagementForm/FormVariants/EndorseBeneficiary/EndorseBeneficiary.tsx [hunk 6]
```diff
@@ -49,22 +50,22 @@ export const EndorseBeneficiaryForm = ({
           beneficiaryAddress: newBeneficiary,
         })
       );
+      onBack(); // onSetFormAction to none
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/FormVariants/EndorseBeneficiary/EndorseBeneficiary.tsx]: !!

File: src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.tsx [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  const { showOverlay } = useContext(OverlayContext);
+
+  useEffect(() => {
+    if (isConfirmed) {
+      showOverlay(
+        showDocumentTransferMessage(MessageTitle.NOMINATE_BENEFICIARY_HOLDER_SUCCESS, {
+          isSuccess: true,
+        })
+      );
+      onBack(); // onSetFormAction to none
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/FormVariants/NominateBeneficiaryHolder/NominateBeneficiaryHolder.tsx]: !!

File: src/components/AssetManagementPanel/AssetManagementForm/FormVariants/SurrenderForm/SurrenderForm.tsx [hunk 8]
```diff
@@ -41,15 +41,14 @@ export const SurrenderForm = ({
   useEffect(() => {
     if (isConfirmed) {
       showOverlay(showDocumentTransferMessage(MessageTitle.SURRENDER_DOCUMENT_SUCCESS, { isSuccess: true }));
+      onBack(); // onSetFormAction to none
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/FormVariants/SurrenderForm/SurrenderForm.tsx]: !!

File: src/components/AssetManagementPanel/AssetManagementForm/FormVariants/TransferHolderForm/TransferHolderForm.tsx [hunk 9]
```diff
@@ -45,22 +46,22 @@ export const TransferHolderForm = ({
       showOverlay(
         showDocumentTransferMessage(MessageTitle.TRANSFER_HOLDER_SUCCESS, { isSuccess: true, holderAddress: newHolder })
       );
+      onBack(); // onSetFormAction to none
```
Comment (yehjxraymond) [inline src/components/AssetManagementPanel/AssetManagementForm/FormVariants/TransferHolderForm/TransferHolderForm.tsx]: !!

File: src/components/UI/Overlay/OverlayContent/DocumentTransferMessage.tsx [hunk 10]
```diff
@@ -138,6 +139,14 @@ export const MessageHolderSuccess = ({ address }: MessageProps) => {
   );
 };
 
+export const MessageNominateBeneficiaryHolderSuccess = () => {
+  return (
+    <>
```
Comment (yehjxraymond) [inline src/components/UI/Overlay/OverlayContent/DocumentTransferMessage.tsx]: !!

File: src/index.js [hunk 11]
```diff
@@ -15,17 +15,17 @@ const history = createBrowserHistory();
...
-    <OverlayContextProvider>
-      <ProviderContextProvider>
-        <Provider store={store}>
-          <TokenInformationContextProvider>
-            <ConnectedRouter history={history}>
+    <Provider store={store}>
+      <ConnectedRouter history={history}>
+        <OverlayContextProvider>
+          <ProviderContextProvider>
+            <TokenInformationContextProvider>
```
Comment (yehjxraymond) [inline src/index.js]: Can explain the reordering?

Comment (simboonlong) [reply src/index.js]: wanted to access useLocation within context, will be reverting this as per discussed

**Review by yehjxraymond — APPROVED** [review]

**Pull request merged**